src/lobpcg/eig.rs

```diff
@@ -0,0 +1,268 @@
+    /// distribution of the eigenvalues. It requires prior knowledge of the problem.
+    pub fn precondition_with(mut self, preconditioner: Array2<A>) -> Self {
+        self.preconditioner = Some(preconditioner);
+
+        self
+    }
+
+    /// Calculate the eigenvalue decomposition
+    ///
+    /// # Parameters
+    ///
+    ///  * `num`: number of eigenvalues ordered by magnitude
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use ndarray::{arr1, Array2};
+    /// use ndarray_linalg_rs::lobpcg::{TruncatedEig, TruncatedOrder};
+    ///
+    /// let diag = arr1(&[1., 2., 3., 4., 5.]);
+    /// let a = Array2::from_diag(&diag);
+    ///
+    /// let eig = TruncatedEig::new(a, TruncatedOrder::Largest)
+    ///    .precision(1e-5)
+    ///    .maxiter(500);
+    ///
+    /// let res = eig.decompose(3);
+    /// ```
+    pub fn decompose(&self, num: usize) -> LobpcgResult<A> {
+        let x: Array2<f64> = random((self.problem.len_of(Axis(0)), num));
+        let x = x.mapv(|x| NumCast::from(x).unwrap());
+
+        if let Some(ref preconditioner) = self.preconditioner {
+            lobpcg(
+                |y| self.problem.dot(&y),
+                x,
+                |mut y| y.assign(&preconditioner.dot(&y)),
+                self.constraints.clone(),
+                self.precision,
+                self.maxiter,
+                self.order.clone(),
+            )
+        } else {
+            lobpcg(
+                |y| self.problem.dot(&y),
+                x,
+                |_| {},
+                self.constraints.clone(),
+                self.precision,
+                self.maxiter,
+                self.order.clone(),
+            )
+        }
+    }
+}
+
+impl<A: Float + NdFloat + PartialOrd + Default + Sum> IntoIterator for TruncatedEig<A> {
+    type Item = (Array1<A>, Array2<A>);
+    type IntoIter = TruncatedEigIterator<A>;
+
+    fn into_iter(self) -> TruncatedEigIterator<A> {
+        TruncatedEigIterator {
+            step_size: 1,
+            remaining: self.problem.len_of(Axis(0)),
+            eig: self,
+        }
+    }
+}
+
+/// Truncated eigenproblem iterator
+///
+/// This wraps a truncated eigenproblem and provides an iterator where each step yields a new
+/// eigenvalue/vector pair. Useful for generating pairs until a certain condition is met.
+///
+/// # Example
+///
+/// ```rust
+/// use ndarray::{arr1, Array2};
+/// use ndarray_linalg_rs::lobpcg::{TruncatedEig, TruncatedOrder};
+///
+/// let diag = arr1(&[1., 2., 3., 4., 5.]);
+/// let a = Array2::from_diag(&diag);
+///
+/// let teig = TruncatedEig::new(a, TruncatedOrder::Largest)
+///     .precision(1e-5)
+///     .maxiter(500);
+///
+/// // solve eigenproblem until eigenvalues get smaller than 0.5
+/// let res = teig.into_iter()
+///     .take_while(|x| x.0[0] > 0.5)
+///     .flat_map(|x| x.0.to_vec())
+///     .collect::<Vec<_>>();
+/// ```
+pub struct TruncatedEigIterator<A: NdFloat> {
+    step_size: usize,
```

Comment:
Right now there's no way in the API to set `step_size` to anything other than 1. Either replace the `IntoIterator` impl with an inherent method `into_iter(self, step_size: usize)` that allows setting step size, or remove this field altogether and always iterate one step at a time. Note that the single-step iterator will yield `(A, Array1<A>)` instead.

Reply:
I added a constructor to `TruncatedEigIterator` to variate the step size

Reply:
Can you add a method called `into_iter_step_by` to `TruncatedEig` that calls the new constructor?

Reply:
added in latest commit